### ISSUE TITLE
EC2 Deploy waits indefinitely

### ISSUE THREAD

**Issue by digitalsanctum** [hashicorp/waypoint#1386]
**Describe the bug**
I'm attempting to use the aws-ec2 plugin to deploy a single instance to a non-default VPC. The AWS I'm working in doesn't have a default VPC. Everything seems to work but CLI waits indefinitely with the message: "Waiting for the first instance to start..."

**Steps to Reproduce**

I'm running the following via CLI:
```
waypoint up -app ec2-web -vvv
```

Here's what my waypoint.hcl app looks like:
```
app "ec2-web" {
  build {
    use "aws-ami" {
      region = "us-west-2"
      name = "ubuntu/images/hvm-ssd/ubuntu-focal-20.04-arm64-server-20210315"
    }
  }

  deploy {
    use "aws-ec2" {
      count {
        desired = 1
        min = 1
        max = 2
      }
      instance_type = "t3.nano"
      key = "beefcake"
      region = "us-west-2"
      service_port = 8080
      subnet = "subnet-89ef61d3"
//      security_groups = ["sg-37d22f44"]
    }
  }

  release {
    use "aws-alb" {
      count = 1
      subnets = ["subnet-89ef61d3"]
    }
  }
}
```

**Expected behavior**
An EC2 instance is deployed.

**Waypoint Platform Versions**
Additional version and platform information to help triage the issue if
applicable:

* Waypoint CLI Version: 0.3.1
* Waypoint Server Platform and Version: docker
* Waypoint Plugin: aws-ec2

**Additional context**
I've confirmed via AWS console that certain things are getting created such as:
* Auto Scaling Group
* Security Group
* Launch Configuration
* Target Group

The VPC in question has 3 subnets of which just 1 is public. That subnet is specified above in the deploy and release stanzas.

Any guidance would be greatly appreciated.

Thanks!

**Comment by briancain** [comment]
Hey there @digitalsanctum ! Are you able to provide us with a debug log in a [gist](https://gist.github.com/) from when you ran `waypoint up -app ec2-web -vvv`? Thank you!

**Comment by digitalsanctum** [comment]
@briancain Sure thing. Here's the gist: https://gist.github.com/digitalsanctum/7578e9940d2ef62385d7c84a02840d9d

**Comment by digitalsanctum** [comment]
Some additional data points:

- If I comment the subnet in the deploy stanza, I get an error: `ValidationError: A VPC ID must be specified` but I see no way to set it from the docs.
- After closer inspection, the docs mention "default VPC" for subnet parameters. In my case, I'm not working with the default VPC but it *is* the only VPC in the account/region. Does this mean that the aws-ec2 plugin will only work with default AWS VPCs?

**Comment by briancain** [comment]
> Some additional data points:
> 
> * If I comment the subnet in the deploy stanza, I get an error: `ValidationError: A VPC ID must be specified` but I see no way to set it from the docs.

This is correct, I don't see a way to set a VPC id with the EC2 plugin. Waypoint obtains the VPC ID through the configured subnet and the aws sdk. It does look like we should be setting a default one though if no subnet is specified 🤔 So perhaps this is a bug, I think that error is coming directly from the aws sdk.

> * After closer inspection, the docs mention "default VPC" for subnet parameters. In my case, I'm not working with the default VPC but it _is_ the only VPC in the account/region. Does this mean that the aws-ec2 plugin will only work with default AWS VPCs?

These docs could be improved here, but I believe what it's saying there is if you don't provide a subnet, it will use a default from the aws sdk which includes a default associated VPC ID. If you specify a different subnet, waypoint _should_ be grabbing the VPC ID associated with that subnet.

I'm guessing the timeout issue here is around the subnet and VPC ID. We'll take a look and try to reproduce in our end. Thanks for all of the extra debugging information! ❤️ 🙏🏻

**Comment by catsby** [comment]
Hey @digitalsanctum  - thanks again for opening this issue. I've been looking into it and have found a few things. 

I reproduced the issue with your config file (thanks for that!), and I did some debugging to find out which exact ASG was created and I looked at it in the AWS console. The first thing I noticed was under `Activity`, it looks like the AMI specified is incompatible with `t3.nano` instance type:

<img width="1530" alt="Screen Shot 2021-05-07 at 11 52 42" src="https://user-images.githubusercontent.com/61721/117483164-fe9b2600-af2a-11eb-9828-2212e1bfaf20.png">

Unfortunately our logic doesn't surface that, it just loops waiting for instances (which in this case, will never launch):

https://github.com/hashicorp/waypoint/blob/fff8248c1ee1aaddb93c00a3c15a948512aecf0b/builtin/aws/ec2/platform.go#L246-L263

I'm working on some logic that will check the [ASG Scaling Activities](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_DescribeScalingActivities.html) and surface any errors due to failure/cancellation, but I thought I'd post this in the event you had not discovered the image / instance type incompatibility yourself yet. 

As far as your question here:

> Does this mean that the aws-ec2 plugin will only work with default AWS VPCs?

I believe this is correct, **if** users do not specify the `subnet` as you originally had. Personally I think that if a user is not using the default VPC, then requiring them set the `subnet` is reasonable (the VPC then being gotten from the subnet). Alternatively we could offer a `vpc_id` field as well. I'm curious if you have thoughts on the usage there?

Thanks again for reporting this, I hope to have a patch that at leasts errors and reports the problem soon

**Comment by catsby** [comment]
Hey @digitalsanctum - I merged #1447 which will check the autoscaling activity and exit the process if the ASG reports a failed or cancelled activity, instead of us just polling indefinitely. Please let us know if there's anything else here!
